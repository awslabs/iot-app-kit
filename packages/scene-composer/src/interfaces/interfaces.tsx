/**
 * This file contains public interfaces of the scene content the SceneComposer exposes.
 * It is different from the serialization format from the GLTF Toolkit package.
 * These interfaces are used to interact with the SceneComposer instance at
 * runtime, thus the data structure is slightly different from the GLTF Toolkit.
 */

import { ReactElement } from 'react';

import { ISceneNodeInternal } from '../store/Store';

import { ISceneComponent, KnownComponentType } from './components';
import { IRuleBasedMap, ITransform, ITransformConstraint, Vector3 } from './dataTypes';

/************************************************
 * Scene document, entity, component models
 ************************************************/

export interface ISceneDocumentSnapshot {
  /**
   * Serialize the current document snapshot into a document string with the
   * given format version.
   *
   * Note that this interface is not optimized for very large scene, which
   * we may introduce new APIs later to use stream API to handle the output.
   *
   * TODO: do we want to allow updating the document version?
   * [Option 1]: Externalize the document version instea of keeping the doucment version in the document itself.
   * [Option 2]: Allow passing in a parameter to set the document version.
   * [Option 3]: Automatically increment the document version by 1 based on the original input document version.
   *
   * We'll go with option 1 for now for simplicity.
   *
   * TODO: what should we return when the scene is empty?
   * [Option 1]: Return a falsy value, such as empty string.
   * [Option 2]: Return a valid scene with no entity.
   *
   * We'll go with option 2 for now.
   *
   * @param specVersion - version of the document format
   */
  serialize(specVersion: string, ...stringifyArgs): string;
}

export interface ISceneDocument {
  nodeMap: Record<string, ISceneNode>;
  ruleMap: Record<string, IRuleBasedMap>;
  rootNodeRefs: string[];
  unit?: string;
  version: string;
  properties?: Partial<Record<KnownSceneProperty, any>>;
  specVersion?: string;
}

export interface ISceneNode {
  // ref is optional when used in APIs to create nodes
  ref?: string;
  name?: string;
  transform?: ITransform;
  transformConstraint?: ITransformConstraint;
  components?: ISceneComponent[];
  parentRef?: string;
  childRefs?: string[];
  properties?: Partial<Record<'alwaysVisible' | 'matterportId' | 'hiddenWhileImmersive', boolean | string>>;
}

export enum KnownSceneProperty {
  BaseUrl = 'baseUrl',
  EnvironmentPreset = 'environmentPreset',
  DataBindingConfig = 'dataBindingConfig',
  ComponentSettings = 'componentSettings',
  MatterportModelId = 'matterportModelId',
  LayerIds = 'layerIds',
  SceneRootEntityId = 'sceneRootEntityId',
  TagCustomColors = 'tagCustomColors',
  FogSettings = 'fogSettings',
}

/************************************************
 * Scene Composer Control APIs
 ************************************************/

export type CameraControlMode = 'transition' | 'teleport';
export type CameraControlsType = 'orbit' | 'pan' | 'immersive' | 'pointerLock';
export type TransformControlMode = 'translate' | 'rotate' | 'scale';
export enum CameraViewAxisValues {
  Front = 'front',
  Back = 'back',
  Left = 'left',
  Right = 'right',
  Top = 'top',
  Bottom = 'bottom',
}

// an object ref or a precise camera setting
export type FixedCameraTarget = { position: Vector3; target: Vector3 };
export type NamedCameraTarget = string;
export type CameraTarget = NamedCameraTarget | FixedCameraTarget;

export type URIModifier = (uri: string) => string;

export type CreateErrorViewCallback = (error: Error) => ReactElement;

/************************************************
 * Scene Namespace
 ************************************************/

const IotTwinMakerNamespace = 'iottwinmaker.common';
export const IotTwinMakerNamespaceSeparator = ':';
export const IotTwinMakerIconNamespace = `${IotTwinMakerNamespace}.icon`;
export const IotTwinMakerColorNamespace = `${IotTwinMakerNamespace}.color`;
export const IotTwinMakerNumberNamespace = `${IotTwinMakerNamespace}.number`;
export const IotTwinMakerOpacityNamespace = `${IotTwinMakerNamespace}.opacity`;
/************************************************
 * Scene Resource
 ************************************************/

export enum SceneResourceType {
  Icon = 'Icon',
  Color = 'Color',
  Number = 'Number',
  Opacity = 'Opacity',
}

export interface SceneResourceInfo {
  type: SceneResourceType;
  value: string;
}

export interface AddingWidgetInfo {
  type?: KnownComponentType;
  node: ISceneNodeInternal;
}

export interface MeshStyle {
  color?: string | number;
  opacity?: number;
  transparent?: boolean;
}

export interface StyleTarget {
  dataBindingContext: unknown;
  style: MeshStyle;
}

export interface IFogSettings {
  colorHex: number;
  near: number;
  far: number;
}
