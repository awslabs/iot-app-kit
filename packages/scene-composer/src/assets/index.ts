export * from './anchors';
export * from './viewpoints';
