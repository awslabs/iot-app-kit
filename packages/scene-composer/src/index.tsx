export * from './interfaces';

export * from './common/errors';

export * from './components/SceneComposerInternal';

export * from './SceneViewer';

export type { Scene } from './models/SceneModels';

// Augmentation related interfaces
export * from './augmentations/three';

export * from './common/GlobalSettings';

export * from './utils/dataFrameLabelUtils';

export * from './utils/dataBindingTemplateUtils';

export * from './common/constants';
