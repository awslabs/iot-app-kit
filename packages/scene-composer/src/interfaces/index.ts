export * from './assets';
export * from './interfaces';
export * from './components';
export * from './dataTypes';
export * from './dataBinding';
export * from './metricRecorder';
export * from './feature';
export * from './sceneViewer';
export * from './componentSettings';
