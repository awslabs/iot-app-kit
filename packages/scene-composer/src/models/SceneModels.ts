/**
 * This file contains the model of the scene file stored in database.
 * Changes to this file should be backward compatible.
 */

export type KeyValuePair = { [key: string]: unknown };
export type UUID = string;
export type Vector3 = [number, number, number];
export type Vector2 = [number, number];
export type Uri = string;

export type DistanceUnit =
  | 'millimeters'
  | 'centimeters'
  | 'decimeters'
  | 'meters'
  | 'kilometers'
  | 'inches'
  | 'feet'
  | 'yards'
  | 'miles';

export type Color = number | string;

export interface GeoLocation {
  longitude: number;
  latitude: number;
  altitude?: number;
}

export interface RuleStatement {
  expression: string;
  target: string;
}

export interface RuleBasedMap {
  statements: Array<RuleStatement>;
}

export interface ValueDataBinding {
  dataBindingContext: unknown;
}

export interface NavLink {
  destination?: string;
  params?: Record<string, string>;
}

export interface Scene {
  specVersion: string;
  version: string;
  unit?: DistanceUnit;
  nodes: Array<Node>;
  rootNodeIndexes: Array<number>;
  properties?: KeyValuePair;
  rules?: Record<string, RuleBasedMap>;
  cameras?: Array<Camera>;
  defaultCameraIndex?: number;
}

export enum CameraType {
  Perspective = 'Perspective',
  Orthographic = 'Orthographic',
}

export interface Camera {
  cameraType: CameraType;
  fov?: number;
  near?: number;
  far?: number;
}

export enum LightType {
  Ambient = 'Ambient',
  Directional = 'Directional',
  Hemisphere = 'Hemisphere',
  Point = 'Point',
}

export interface Transform {
  position: Vector3;
  rotation?: Vector3; // euler angle
  scale?: Vector3;
}

export interface TransformConstraint {
  snapToFloor?: boolean;
}

export interface Node {
  name?: string;
  transform: Transform;
  transformConstraint?: TransformConstraint;
  components?: Array<Component.IComponent>;
  children?: Array<number>;
  properties?: Record<string, any>;
}

export enum ModelType {
  GLB = 'GLB',
  GLTF = 'GLTF',
  Tiles3D = 'Tiles3D', // 3D Tiles
  Environment = 'Environment', // Environment for external sources
}

export namespace Component {
  export enum Type {
    Camera = 'Camera',
    Light = 'Light',

    ModelRef = 'ModelRef',
    SubModelRef = 'SubModelRef',
    Tag = 'Tag',
    ModelShader = 'ModelShader',
    OpacityFilter = 'OpacityFilter',
    MotionIndicator = 'MotionIndicator',
    Space = 'Space',
    DataOverlay = 'DataOverlay',
  }

  export interface IComponent {
    type: Type | string;
  }

  export interface IDataBindingRuleMap {
    valueDataBinding?: ValueDataBinding;
    ruleBasedMapId?: string;
  }

  export interface Space extends IComponent {
    spaceId: string;
    parentSpaceEntityId: string;
    bounds: Vector3;
    boundsOffset: Vector3;
  }

  export interface ModelRef extends IComponent {
    uri: Uri;
    modelType: ModelType;
    unitOfMeasure?: DistanceUnit;
    localScale?: Vector3;
    castShadow?: boolean;
    receiveShadow?: boolean;
  }

  export interface SubModelRef extends IComponent {
    parentRef: string;
    selector: string | number;
  }

  export interface Camera extends IComponent {
    cameraIndex: number;
  }

  export interface Tag extends IComponent, IDataBindingRuleMap {
    icon?: string;
    navLink?: NavLink;
    offset?: Vector3;
  }

  export interface ModelShader extends IComponent, IDataBindingRuleMap {}
  export interface OpacityFilter extends IComponent, IDataBindingRuleMap {}

  export interface IMotionIndicatorConfig {
    numOfRepeatInY: number;
    backgroundColorOpacity: number;
    defaultBackgroundColor?: Color;
    defaultForegroundColor?: Color;
    defaultSpeed?: number;
  }
  export type ILinearPlaneMotionIndicatorConfig = IMotionIndicatorConfig;

  export type ILinearCylinderMotionIndicatorConfig = IMotionIndicatorConfig;

  export type ICircularCylinderMotionIndicatorConfig = IMotionIndicatorConfig;

  export enum MotionIndicatorDataBindingName {
    Speed = 'speed',
    ForegroundColor = 'foregroundColor',
    BackgroundColor = 'backgroundColor',
  }

  export enum MotionIndicatorShape {
    LinearPlane = 'LinearPlane',
    LinearCylinder = 'LinearCylinder',
    CircularCylinder = 'CircularCylinder',
  }

  export interface MotionIndicator extends IComponent {
    valueDataBindings: { [key in MotionIndicatorDataBindingName]?: IDataBindingRuleMap };

    shape: MotionIndicatorShape;
    config:
      | ILinearPlaneMotionIndicatorConfig
      | ILinearCylinderMotionIndicatorConfig
      | ICircularCylinderMotionIndicatorConfig;
  }

  // Data Overlay
  export enum DataOverlayRowType {
    Markdown = 'Markdown',
  }
  export interface DataOverlayRow {
    rowType: DataOverlayRowType;
  }
  export interface DataOverlayMarkdownRow extends DataOverlayRow {
    content: string;
  }
  export enum DataOverlaySubType {
    TextAnnotation = 'TextAnnotation',
    OverlayPanel = 'OverlayPanel',
  }
  // TODO: Not supported in milestone 1
  export interface OverlayPanelConfig {
    isPinned: boolean;
  }

  export interface ValueDataBindingNamedMap {
    bindingName: string;
    valueDataBinding?: ValueDataBinding;
  }

  export interface DataOverlay extends IComponent {
    subType: DataOverlaySubType;
    dataRows: Array<DataOverlayMarkdownRow>;

    // TODO: Not supported in milestone 1
    // config?: OverlayPanelConfig;

    valueDataBindings: ValueDataBindingNamedMap[];
  }

  export interface ILightShadowSettings {
    shadowBias?: number;
    shadowCameraLeft?: number;
    shadowCameraRight?: number;
    shadowCameraTop?: number;
    shadowCameraBottom?: number;
    shadowMapSizeWidth?: number;
    shadowMapSizeHeight?: number;
  }

  export interface IDirectionalLightSettings {
    color: Color;
    intensity: number;

    castShadow: boolean;
    shadow?: ILightShadowSettings;
  }

  export interface IPointLightSettings {
    color: Color;
    intensity: number;
    // Maximum range of the light. Default is 0 (no limit).
    distance?: number;
    // The amount the light dims along the distance of the light. Default is 1.
    // For physically correct lighting, set this to 2.
    decay?: number;

    castShadow: boolean;
    shadow?: ILightShadowSettings;
  }

  export interface IAmbientLightSettings {
    color: Color;
    intensity: number;
  }

  export interface IHemisphereLightSettings {
    color: Color;
    groundColor: Color;
    intensity: number;
  }

  export interface Light extends IComponent {
    lightType: LightType;
    lightSettings: IDirectionalLightSettings | IAmbientLightSettings | IHemisphereLightSettings;
  }
}
