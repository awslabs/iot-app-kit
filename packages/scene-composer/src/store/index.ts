export * from './StoreOperations';
export * from './Store';
