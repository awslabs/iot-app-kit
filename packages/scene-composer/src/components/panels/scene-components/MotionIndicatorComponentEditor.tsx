import React, { useCallback, useContext } from 'react';
import { FormField, Select, SpaceBetween } from '@cloudscape-design/components';
import { useIntl, defineMessages } from 'react-intl';

import { IComponentEditorProps } from '../ComponentEditor';
import { IMotionIndicatorComponentInternal, ISceneComponentInternal, useStore } from '../../../store';
import { sceneComposerIdContext } from '../../../common/sceneComposerIdContext';
import { Component } from '../../../models/SceneModels';
import { Divider } from '../../Divider';

import AppearanceEditor from './motion-indicator/AppearanceEditor';
import { SpeedEditor } from './motion-indicator/SpeedEditor';

export type IMotionIndicatorComponentEditorProps = IComponentEditorProps;

export const MotionIndicatorComponentEditor: React.FC<IMotionIndicatorComponentEditorProps> = ({
  node,
  component,
}: IMotionIndicatorComponentEditorProps) => {
  const sceneComposerId = useContext(sceneComposerIdContext);
  const updateComponentInternal = useStore(sceneComposerId)((state) => state.updateComponentInternal);
  const motionIndicatorComponent = component as IMotionIndicatorComponentInternal;
  const selectedShape = motionIndicatorComponent.shape;
  const intl = useIntl();

  const i18nMotionIndicatorShapeStrings = defineMessages({
    LinearPlane: {
      defaultMessage: 'Linear Plane',
      description: 'Motion Indicator Shape in a dropdown menu',
    },
    LinearCylinder: {
      defaultMessage: 'Linear Cylinder',
      description: 'Motion Indicator Shape in a dropdown menu',
    },
    CircularCylinder: {
      defaultMessage: 'Circular Cylinder',
      description: 'Motion Indicator Shape in a dropdown menu',
    },
  });

  const onUpdateCallback = useCallback(
    (componentPartial: any, replace?: boolean) => {
      const componentPartialWithRef: ISceneComponentInternal = { ref: component.ref, ...componentPartial };
      updateComponentInternal(node.ref, componentPartialWithRef, replace);
    },
    [node.ref, component.ref],
  );

  const shapeOptions = Object.values(Component.MotionIndicatorShape).map((shape) => ({
    label: intl.formatMessage(i18nMotionIndicatorShapeStrings[shape]) || shape,
    value: shape,
  }));

  return (
    <SpaceBetween size='s'>
      <FormField label={intl.formatMessage({ defaultMessage: 'Indicator shape', description: 'FormField label' })}>
        <Select
          data-testid='motion-indicator-shape-select'
          selectedOption={
            selectedShape
              ? {
                  label: intl.formatMessage(i18nMotionIndicatorShapeStrings[selectedShape]) || selectedShape,
                  value: selectedShape,
                }
              : null
          }
          onChange={(e) => {
            const shape = e.detail.selectedOption.value;
            const updatedComponent = { ...motionIndicatorComponent, shape };
            onUpdateCallback(updatedComponent, true);
          }}
          options={shapeOptions}
          selectedAriaLabel={intl.formatMessage({
            defaultMessage: 'Selected',
            description:
              'Specifies the localized string that describes an option as being selected. This is required to provide a good screen reader experience',
          })}
          placeholder={intl.formatMessage({ defaultMessage: 'Choose a shape', description: 'Placeholder' })}
        />
      </FormField>

      <Divider />

      <SpeedEditor component={motionIndicatorComponent} onUpdateCallback={onUpdateCallback} />

      <Divider />
      <AppearanceEditor
        component={motionIndicatorComponent}
        onUpdateCallback={onUpdateCallback}
        scale={node.transform.scale}
      />
    </SpaceBetween>
  );
};
