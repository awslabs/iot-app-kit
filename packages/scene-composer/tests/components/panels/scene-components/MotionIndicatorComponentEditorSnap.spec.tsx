import React from 'react';
import { render } from '@testing-library/react';

import { MotionIndicatorComponentEditor } from '../../../../src/components/panels/scene-components/MotionIndicatorComponentEditor';
import { IMotionIndicatorComponentInternal, useStore } from '../../../../src/store';
import { KnownComponentType } from '../../../../src/interfaces';
import { Component } from '../../../../src/models/SceneModels';

import { mockNode, mockComponent } from './MockComponents';

jest.mock('../../../../src/components/panels/scene-components/motion-indicator/SpeedEditor', () => {
  const originalModule = jest.requireActual(
    '../../../../src/components/panels/scene-components/motion-indicator/SpeedEditor',
  );
  return {
    ...originalModule,
    SpeedEditor: (...props: any[]) => <div id='SpeedEditor'>{JSON.stringify(props)}</div>,
  };
});

jest.mock('../../../../src/components/panels/scene-components/motion-indicator/AppearanceEditor', () => {
  const originalModule = jest.requireActual(
    '../../../../src/components/panels/scene-components/motion-indicator/AppearanceEditor',
  );
  return {
    ...originalModule,
    AppearanceEditor: (...props: any[]) => <div id='AppearanceEditor'>{JSON.stringify(props)}</div>,
  };
});

const updateComponentInternalFn = jest.fn();

const baseState = {
  updateComponentInternal: updateComponentInternalFn,
};

describe('MotionIndicatorComponentEditor', () => {
  const component: IMotionIndicatorComponentInternal = {
    ...mockComponent,
    type: KnownComponentType.MotionIndicator,
    shape: Component.MotionIndicatorShape.LinearPlane,
    valueDataBindings: {},
    config: {
      numOfRepeatInY: 4,
      backgroundColorOpacity: 0.5,
    },
  };

  it('should render correctly', () => {
    useStore('default').setState(baseState);

    const { container } = render(
      <MotionIndicatorComponentEditor node={{ ...mockNode, components: [component] }} component={component} />,
    );
    expect(container).toMatchSnapshot();
  });
});
