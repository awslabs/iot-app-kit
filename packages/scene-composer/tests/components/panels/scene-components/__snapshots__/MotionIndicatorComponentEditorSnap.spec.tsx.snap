// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MotionIndicatorComponentEditor should render correctly 1`] = `
.c0 {
  height: 1px;
  border-width: 0px;
  background-color: colorBorderDividerDefault;
}

<div>
  <spacebetween>
    <formfield
      label="Indicator shape"
    >
      <select
        data-testid="motion-indicator-shape-select"
        options="[object Object],[object Object],[object Object]"
        placeholder="Choose a shape"
        selectedarialabel="Selected"
        selectedoption="[object Object]"
      />
    </formfield>
    <hr
      class="c0"
    />
    <div
      id="SpeedEditor"
    >
      [{"component":{"ref":"componentRef1","type":"MotionIndicator","uri":"componentUri","modelType":"componentModelType","shape":"LinearPlane","valueDataBindings":{},"config":{"numOfRepeatInY":4,"backgroundColorOpacity":0.5}}},{}]
    </div>
    <hr
      class="c0"
    />
    <div
      id="AppearanceEditor"
    >
      [{"component":{"ref":"componentRef1","type":"MotionIndicator","uri":"componentUri","modelType":"componentModelType","shape":"LinearPlane","valueDataBindings":{},"config":{"numOfRepeatInY":4,"backgroundColorOpacity":0.5}},"scale":[0,0,0]},{}]
    </div>
  </spacebetween>
</div>
`;
